core/src/message/context.rs

```diff
@@ -313,3 +315,215 @@ impl MessageContext {
+            incoming_message,
+            ids::ProgramId::from(INCOMING_MESSAGE_ID),
+            None,
+        );
+
+        // Checking that the initial parameters of the context match the passed constants
+        assert_eq!(context.current().id(), MessageId::from(INCOMING_MESSAGE_ID));
+        assert!(context.store.reply.is_none());
+        assert!(context.outcome.reply.is_none());
+
+        // Creating a reply packet
+        let reply_packet = ReplyPacket::new(vec![0, 0], 0);
+
+        // Checking that we are able to initialize reply
+        assert!(context.reply_push(&[1, 2, 3]).is_ok());
+
+        // Setting reply message and making sure the operation was successful
+        assert!(context.reply_commit(reply_packet.clone()).is_ok());
+
+        // Checking that the `ReplyMessage` matches the passed one
+        assert_eq!(
+            context.outcome.reply.as_ref().unwrap().payload().to_vec(),
+            vec![1, 2, 3, 0, 0],
+        );
+
+        // Checking that repeated call `reply_push(...)` returns error and does not do anything
+        assert!(context.reply_push(&[1]).is_err());
+        assert_eq!(
+            context.outcome.reply.as_ref().unwrap().payload().to_vec(),
+            vec![1, 2, 3, 0, 0],
+        );
+
+        // Checking that repeated call `reply_commit(...)` returns error and does not
+        assert!(context.reply_commit(reply_packet).is_err());
+
+        // Checking that at this point vector of outgoing messages is empty
+        assert!(context.outcome.handle.is_empty());
+
+        // Creating an expected handle for a future initialized message
+        let expected_handle = 0;
+
+        // Initializing message and compare its handle with expected one
+        assert_eq!(
+            context.send_init().expect("Error initializing new message"),
+            expected_handle
+        );
+
+        // And checking that it is not formed
+        assert!(context
+            .store
+            .outgoing
+            .get(&expected_handle)
+            .expect("This key should be")
+            .is_some());
+
+        // Checking that we are able to push payload for the
+        // message that we have not committed yet
+        assert!(context.send_push(expected_handle, &[5, 7]).is_ok());
+        assert!(context.send_push(expected_handle, &[9]).is_ok());
+
+        // Creating an outgoing packet to commit sending by parts
+        let commit_packet = HandlePacket::default();
+
+        // Checking if commit is successful
+        assert!(context.send_commit(expected_handle, commit_packet).is_ok());
+
+        // Checking that we are **NOT** able to push payload for the message or
+        // commit it if we already committed it or directly pushed before
+        assert!(context.send_push(0, &[5, 7]).is_err());
+        assert!(context.send_push(expected_handle, &[5, 7]).is_err());
+        assert!(context.send_commit(0, HandlePacket::default()).is_err());
```

Comment:
why `expected_handle` isn't used in all calls?

Reply:
The lines 495 and 497 duplicates the lines below. This can be a check that there will be an error no matter how many times we repeat the action. But it's weird, so I deleted them.